vllm/model_executor/layers/quantization/awq.py

```diff
@@ -86,54 +107,102 @@ def create_weights(self, input_size_per_partition: int,
                 "The input size is not aligned with the quantized "
                 "weight shape. This can be caused by too large "
                 "tensor parallel size.")
-        if output_size_per_partition % self.quant_config.pack_factor != 0:
+        if output_size_per_partition % self.quant_config.pack_factor_int32 != 0:
             raise ValueError(
                 "The output size is not aligned with the quantized "
                 "weight shape. This can be caused by too large "
                 "tensor parallel size.")
 
-        qweight = Parameter(
-            torch.empty(
-                input_size_per_partition,
-                output_size_per_partition // self.quant_config.pack_factor,
-                dtype=torch.int32,
-            ),
-            requires_grad=False,
-        )
-        set_weight_attrs(
-            qweight, {
-                "input_dim": 0,
-                "output_dim": 1,
-                "packed_dim": 1,
-                "pack_factor": self.quant_config.pack_factor,
-            })
-        qzeros = Parameter(
-            torch.empty(
-                input_size_per_partition // self.quant_config.group_size,
-                output_size_per_partition // self.quant_config.pack_factor,
-                dtype=torch.int32,
-            ),
-            requires_grad=False,
-        )
-        set_weight_attrs(
-            qzeros, {
+        if self.quant_config:
+            qweight = Parameter(
+                torch.empty(
+                    output_size_per_partition // self.quant_config.interleave,
+                    input_size_per_partition // self.quant_config.pack_factor_int16 * self.quant_config.interleave,
+                    dtype=torch.int16,
+                ),
+                requires_grad=False,
+            )
+            set_weight_attrs(
+                qweight, {
+                    "input_dim": 1,
+                    "output_dim": 0,
+                    "packed_dim": 1,
+                    "pack_factor": self.quant_config.pack_factor_int16,
+                    "interleave": self.quant_config.interleave,
+                })
+            qzeros = Parameter(
+                torch.empty(
+                    calculate_zeros_width(input_size_per_partition, self.quant_config.group_size) * self.quant_config.pack_factor_int32,
+                    output_size_per_partition,
+                    dtype=params_dtype,
+                ),
+                requires_grad=False,
+            )
+            set_weight_attrs(
+                qzeros, {
+                    "input_dim": 0,
+                    "output_dim": 1,
+                    "packed_dim": 0,
+                    "pack_factor": self.quant_config.pack_factor_int32,
```

Comment:
The qzeros.pack_factor and the scales.pack_factor below will cause the qzeros and scales being wrongly loaded. After changing this to group_size, and merging the @robertgshaw2-neuralmagic 's modification, the fast kernel can be successfully run. However, the performance is not good as expected. I wonder if there is still any bugs to fix to make the fast kernel performing well.

Reply:
vLLM is fundamentally not geared for quantized models due to the nature of blocking decoding while prefilling. Chunked prefilling should help. 

However, the new kernels are much faster than the older one’s at decoding. It may not be obvious if you are not testing for this specifically.

Can you upstream your changes in a PR to my repo so then we have a common version that we can test?

Reply:
> vLLM is fundamentally not geared for quantized models due to the nature of blocking decoding while prefilling. Chunked prefilling should help.
> 
> However, the new kernels are much faster than the older one’s at decoding. It may not be obvious if you are not testing for this specifically.
> 
> Can you upstream your changes in a PR to my repo so then we have a common version that we can test?

Of course!